_posts/2023-03-27-group-equivariant-cnn-2/group-equivariant-cnn-2.Rmd

```diff
@@ -0,0 +1,374 @@
+
+Returning to things implementation, it turns out that group actions are encoded as matrices. There is one matrix for each group element. For $C_4$, the so-called *regular* representation is a rotation matrix:
+
+$$
+\begin{bmatrix} \cos(\theta) & -\sin(\theta) \\ \sin(\theta) & \cos(\theta) \end{bmatrix}
+$$
+
+In `gcnn`, the function applying that matrix is `C_4$left_action_on_R2()`. Like its sibling, it is designed to work with batches (of group elements as well as $\mathbb{R}^2$ vectors). Technically, what it does is rotate the grid the image is defined on, and then, re-sample the image. To make this more concrete, that method's code looks about as follows.
+
+Here is a goat.
+
+```{r}
+img_path <- system.file("imgs", "z.jpg", package = "gcnn")
+img <- torchvision::base_loader(img_path) |> torchvision::transform_to_tensor()
+img$permute(c(2, 3, 1)) |> as.array() |> as.raster() |> plot()
+```
+
+```{r, echo=FALSE, eval=TRUE, fig.alt = "A goat sitting comfortably on a meadow."}
+knitr::include_graphics("images/z.jpg")
+```
+
+First, we call `C_4$left_action_on_R2()` to rotate the grid.
+
+```{r}
+# Grid shape is [2, 1024, 1024], for a 2d, 1024 x 1024 image.
+img_grid_R2 <- torch::torch_stack(torch::torch_meshgrid(
+    list(
+      torch::torch_linspace(-1, 1, dim(img)[2]),
+      torch::torch_linspace(-1, 1, dim(img)[3])
+    )
+))
+
+# Transform the image grid with the matrix representation of some group element.
+transformed_grid <- C_4$left_action_on_R2(C_4$inverse(g1)$unsqueeze(1), img_grid_R2)
+
+```
+
+Second, we re-sample the image on the transformed grid. The goat now looks up to the sky.
+
+```{r}
+transformed_img <- torch::nnf_grid_sample(
+  img$unsqueeze(1), transformed_grid,
+  align_corners = TRUE, mode = "bilinear", padding_mode = "zeros"
+)
+
+transformed_img[1,..]$permute(c(2, 3, 1)) |> as.array() |> as.raster() |> plot()
+```
+
+```{r, echo=FALSE, eval=TRUE, fig.alt = "Same goat, rotated up by 90 degrees."}
+knitr::include_graphics("images/z2.jpg")
+```
+
+## Step 2: The lifting convolution
+
+We want to make use of existing, efficient `torch` functionality as much as possible. Concretely, we want to use `nn_conv2d()`. What we need, though, is a convolution kernel that's equivariant not just to translation, but also, to the action of $C_4$. This can be achieved by having one kernel each for every possible rotation.
```

Comment:
```suggestion
We want to make use of existing, efficient `torch` functionality as much as possible. Concretely, we want to use `nn_conv2d()`. What we need, though, is a convolution kernel that's equivariant not just to translation, but also to the action of $C_4$. This can be achieved by having one kernel for each possible rotation.
```